panic: Cannot convert nil datum to EncDatum

Hey,

just recently cockroachdb crashed.

It took a few restarts until it stopped crashing again.

I guess it is related to schema changes, but it doesn't seem to happen always.

```
grep -F '[config]' cockroach-data/logs/cockroach.log
I180501 02:45:45.645510 20686 util/log/clog.go:1001  [config] file created at: 2018/05/01 02:45:45
I180501 02:45:45.645510 20686 util/log/clog.go:1001  [config] running on machine: db-node-1
I180501 02:45:45.645510 20686 util/log/clog.go:1001  [config] binary: CockroachDB CCL v2.0.1 (x86_64-unknown-linux-gnu, built 2018/04/23 18:39:21, go1.10)
I180501 02:45:45.645510 20686 util/log/clog.go:1001  [config] arguments: [cockroach-server/cockroach --cache=25% --max-sql-memory=25% --max-disk-temp-storage=25% --logtostderr=INFO --certs-dir=certs start]
```

```
panic: Cannot convert nil datum to EncDatum

goroutine 86768784 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:137 +0x267
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*memRowContainer).EncRow(0xc43f575900, 0x0, 0xc43c868450, 0xc426759860, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/row_container.go:133 +0x168
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKProcessor).Next(0xc43e10f680, 0xe55f90, 0x2771c80, 0xc45c6ccab0, 0x2791d00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:433 +0x4c7
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x2771bc0, 0xc42f305240, 0x2772a80, 0xc43e10f680, 0x2756f40, 0xc4500438c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:137 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKProcessor).Run(0xc43e10f680, 0xc42d63fb38)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:451 +0x6b
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:470 +0x3a5
```

We have another report of a similar issue from Gitter:

```sql
-- build:               CCL v2.0.1 @ 2018/04/23 18:39:21 (go1.10)
create database test;
set database=test;
create table atable (id bigserial primary key, d text);
insert into atable (d) values ('test1');
select * from atable order by d asc offset 1; -- works
select * from atable order by d asc offset 1 limit 10; -- kills cockroach
```
stack trace
```bash
panic: Cannot convert nil datum to EncDatum

goroutine 1032 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:137 +0x267
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*memRowContainer).EncRow(0xc42a3b2f00, 0x0, 0xc42a3bc2d0, 0xc42ce15020, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/row_container.go:133 +0x168
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKProcessor).Next(0xc4206a4d80, 0xe55f90, 0x2771c80, 0xc42cda6cf0, 0x2791d00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:433 +0x4c7
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x2771bc0, 0xc4292e2980, 0x2772a80, 0xc4206a4d80, 0x2756f40, 0xc42cee0900)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:137 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKProcessor).Run(0xc4206a4d80, 0xc4296967f8)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:451 +0x6b
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:470 +0x3a5
```

--- a/pkg/sql/distsqlrun/sorter.go
+++ b/pkg/sql/distsqlrun/sorter.go
@@ -425,11 +425,11 @@ func (s *sortTopKProcessor) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
 		s.rows.Sort(ctx)
 	}
 
-	if s.closed || s.rows.Len() == 0 {
-		return nil, s.producerMeta(nil /* err */)
-	}
-
 	for {
+		if s.closed || s.rows.Len() == 0 {
+			return nil, s.producerMeta(nil /* err */)
+		}
+
 		row := s.rows.EncRow(0)
 		s.rows.PopFirst()
 

--- a/pkg/sql/logictest/testdata/logic_test/order_by
+++ b/pkg/sql/logictest/testdata/logic_test/order_by
@@ -838,3 +838,16 @@ limit           ·      ·               (block_id, writer_id, block_num, block_
 ·               table  blocks@primary  ·                                                     ·
 ·               spans  ALL             ·                                                     ·
 ·               limit  1               ·                                                     ·
+
+# Regression test for #25197: make sure that order by with an offset works in
+# the top-k sorter.
+
+statement ok
+CREATE TABLE t25197 (id BIGSERIAL PRIMARY KEY, d TEXT)
+
+statement ok
+INSERT INTO t25197 (d) VALUES ('test1')
+
+query TT
+SELECT * FROM t25197 ORDER BY d ASC OFFSET 1 LIMIT 10
+----

